plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs

```diff
@@ -0,0 +1,251 @@
+  , pluginHandlers = mkPluginHandler STextDocumentCodeAction licenseSuggestCodeAction
+  , pluginNotificationHandlers = mconcat
+  [ mkPluginNotificationHandler LSP.STextDocumentDidOpen $
+      \ide vfs _ (DidOpenTextDocumentParams TextDocumentItem{_uri,_version}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocOpened _uri
+        addFileOfInterest ide file Modified{firstOpen=True}
+        restartCabalShakeSession ide vfs file "(opened)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidChange $
+      \ide vfs _ (DidChangeTextDocumentParams VersionedTextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocModified _uri
+        addFileOfInterest ide file Modified{firstOpen=False}
+        restartCabalShakeSession ide vfs file "(changed)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidSave $
+      \ide vfs _ (DidSaveTextDocumentParams TextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocSaved _uri
+        addFileOfInterest ide file OnDisk
+        restartCabalShakeSession ide vfs file "(saved)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidClose $
+      \ide vfs _ (DidCloseTextDocumentParams TextDocumentIdentifier{_uri}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocClosed _uri
+        deleteFileOfInterest ide file
+        restartCabalShakeSession ide vfs file "(closed)"
+  ]
+  }
+  where
+    log' = logWith recorder
+
+    whenUriFile :: Uri -> (NormalizedFilePath -> IO ()) -> IO ()
+    whenUriFile uri act = whenJust (LSP.uriToFilePath uri) $ act . toNormalizedFilePath'
+
+-- | Helper function to restart the shake session, specifically for modifying .cabal files.
+-- No special logic, just group up a bunch of functions you need for the base
+-- Notification Handlers.
+restartCabalShakeSession :: IdeState -> VFS.VFS -> NormalizedFilePath -> String -> IO ()
+restartCabalShakeSession ide vfs file actionMsg = do
+  join $ atomically $ Shake.recordDirtyKeys (shakeExtras ide) GetModificationTime [file]
+  restartShakeSession (shakeExtras ide) (VFSModified vfs) (fromNormalizedFilePath file ++ " " ++ actionMsg) []
+
+-- ----------------------------------------------------------------
+-- Plugin Rules
+-- ----------------------------------------------------------------
+
+data ParseCabal = ParseCabal
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable ParseCabal
+instance NFData   ParseCabal
+
+type instance RuleResult ParseCabal = ()
+
+cabalRules :: Recorder (WithPriority Log) -> Rules ()
+cabalRules recorder = do
+  -- Make sure we initialise the cabal files-of-interest.
+  ofInterestRules recorder
+  -- Rule to produce diagnostics for cabal files.
+  define (cmapWithPrio LogShake recorder) $ \ParseCabal file -> do
+    t <- use GetModificationTime file
+    log' Debug $ LogModificationTime file t
```

Comment:
I'm not sure what this log means. It doesn't seem like anything is being _modified_ here?

Reply:
The intent is to say, that we log the modification time of the file we want to parse. Afaict, this makes sure that our rule depends on the last modification time of that file, and only runs the rule, if the dependency is dirty.

Reply:
Okay, I kind of get that (but write it down!), I just find the log confusing. If it was `LogParsingCabalFile file t` that would make more sense to me.

Reply:
And now it arguably doesn't tell us much, given that we should be relying on GetFileContents?

Reply:
I added a comment, but I am not sure we need more than that